File: src/text_parser.py
# !/usr/bin/env python3

#from lark import Lark, Transformer
from text_grammer_lark import Lark_StandAlone, Transformer

class InputTextTransformer(Transformer):
    def __init__(self, dict_, trie, visit_tokens=True):
        super().__init__(visit_tokens)
        self.dict = dict_
        self.trie = trie
        self.dict.setdefault('\n', ['Newline'])
        self.sep = self.dict.get('SEPARATOR', '\x1F')

    def ensure_list(self, arg):
        if isinstance(arg, list):
            return arg
        else:
            return [arg]

    def break_word(self, string):
        n = len(string)
        start = 0
        while start < n:
            node = self.trie
            word = string[start]
            for pos in range(start, n):
                c = string[pos]
                if c not in node:
                    break
                node = node[c]
                word = node.get('word', word)
            chars = self.dict.get(word, 'Null')
            start += len(word)
            yield from self.ensure_list(chars)

    def start(self, args):
        return args[0]

    def text(self, args):
        ret = []
        buf = []
        for arg in args:
            if arg[0] in ('Space', 'Newline'):
                buf.extend(arg)
            else:
                ret.extend(buf)
                ret.extend(arg)
                buf = []
        # Leave trailing spaces in buf
        return ret

    def WORD(self, tok):
        if schars := self.dict.get(tok, None):
            return self.ensure_list(schars)

        if self.sep in tok:
            words = filter(None, tok.split(self.sep))
            schars = [schar for word in words for schar in self.WORD(word)]
            return schars

        return list(self.break_word(tok))


    def WHITESPACE(self, tok):
        return self.ensure_list(self.dict.get(tok, ['Space']))

#parser = Lark.open(
#    "text_grammer.lark",
#    parser="lalr",
#)
parser = Lark_StandAlone()

def parse_text(text, wdict, trie):
    tree = parser.parse(text)
    return InputTextTransformer(wdict, trie).transform(tree)
if __name__ == "__main__":
    text = """
    あ-い  あい　うえ う-え お
    """
    wdict = {
        'あ': ['A'],
        'い': ['I'],
        'うえ': ['Ue'],
        'う': ['U'],
        'え': ['E'],
    }
    res = parse_text(text, wdict)
    print(res)

File: src/TextParser.js
export class TextParser {
    /**
     * @param {Object} wdict - 辞書データ (sh['dictionary'])
     * @param {Object} trie - Trie木データ (sh['trie'])
     */
    constructor(wdict, trie) {
        this.dict = wdict || {};
        this.trie = trie || {};
        
        // デフォルト値の設定 (Python側の挙動を再現)
        if (!this.dict['\n']) {
            this.dict['\n'] = ['Newline'];
        }
        this.sep = this.dict['SEPARATOR'] || '\x1F';
    }

    ensureList(arg) {
        if (Array.isArray(arg)) return arg;
        return [arg];
    }

    /**
     * Pythonの break_word(string) の最長一致ロジックを再現
     */
    *breakWord(string) {
        const n = string.length;
        let start = 0;
        while (start < n) {
            let node = this.trie;
            let word = string[start];
            
            for (let pos = start; pos < n; pos++) {
                const c = string[pos];
                if (node && c in node) {
                    node = node[c];
                    if (node && node['word']) {
                        word = node['word'];
                    }
                } else {
                    break;
                }
            }
            
            const chars = this.dict[word] || 'Null';
            start += word.length;
            yield* this.ensureList(chars);
        }
    }

    /**
     * 単語（WORD）トークンの処理
     */
    parseWord(tok) {
        if (this.dict[tok]) {
            return this.ensureList(this.dict[tok]);
        }

        if (tok.includes(this.sep)) {
            const words = tok.split(this.sep).filter(Boolean);
            const schars = [];
            for (const word of words) {
                schars.push(...this.parseWord(word));
            }
            return schars;
        }

        return Array.from(this.breakWord(tok));
    }

    /**
     * 空白（WHITESPACE）トークンの処理
     */
    parseWhitespace(tok) {
        return this.ensureList(this.dict[tok] || ['Space']);
    }

    /**
     * Python版 InputTextTransformer.text(args) のバッファ制御を再現
     * 連続する空白・改行が次の通常単語の前に正しくマージされるように処理します
     */
    parse(text) {
        // Larkパーサーの代わりに、正規表現で「非空白の連続（WORD）」と「空白の連続（WHITESPACE）」に分解
        const tokens = text.match(/\s+|\S+/g) || [];
        
        const args = tokens.map(tok => {
            if (/\s/.test(tok)) {
                return this.parseWhitespace(tok);
            } else {
                return this.parseWord(tok);
            }
        });

        const ret = [];
        let buf = [];

        for (const arg of args) {
            if (arg.length > 0 && (arg[0] === 'Space' || arg[0] === 'Newline')) {
                buf.push(...arg);
            } else {
                ret.push(...buf);
                ret.push(...arg);
                buf = [];
            }
        }
        // 末尾に残った空白バッファはPython版の仕様通り捨てる（または必要に応じて結合）
        return ret;
    }
}

/**
 * 外部から簡単に呼び出すためのラップ関数
 */
export function parseText(text, wdict, trie) {
    const parser = new TextParser(wdict, trie);
    return parser.parse(text);
}

